File: src/gestorAplicacion/gestion/Subsidio.java
package gestorAplicacion.gestion;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.lang.Math;
import baseDatos.Serializador;
import gestorAplicacion.gestion.*;
import gestorAplicacion.personas.*;

public class Subsidio {
	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;
	private final float condicion;
	private ArrayList<Estudiante> beneficiarios;
	private int cupos;
	public static ArrayList<Subsidio> listaSubsidio = new ArrayList<Subsidio>();
	private String nombre;

	public Subsidio( String nombre,float condicion, int cupos) {
		super();
		this.condicion = condicion;
		this.cupos = cupos;
		this.nombre = nombre;
		this.beneficiarios = new ArrayList<Estudiante>();
		Subsidio.listaSubsidio.add(this);
		Collections.sort(Estudiante.getListaEstudiantes(), new SortbyIES());
		this.setBeneficiarios();

	}

	public static float calcularIES(Estudiante beneficiario) {
		float puntuacion = (beneficiario.getIngresosFamiliares() * 100) / (beneficiario.getNumeroFamiliares() * 760000);
		// Tomando 760.000 como el ingreso promedio por persona
		int ubicacion = (int) Math.pow(beneficiario.getEstrato(), 2f) - 4;
		if (beneficiario.isColegioPublico()) {
			puntuacion -= 10;
		}
		puntuacion += ubicacion - 100;
		return puntuacion;
	}


	public void setBeneficiarios() {
		ArrayList<Estudiante> estudiantesGeneral = Estudiante.getListaEstudiantes();
		for (Estudiante estudiante : estudiantesGeneral) {
			float result = calcularIES(estudiante);
			boolean cuposDisponibles = cupos > 0;
			if ((result < condicion) && cuposDisponibles) {
				beneficiarios.add(estudiante);
				cupos--;
			} else {
			}
		}

	}
	public float promedioBeneficiarios() {
		float total = 0;
		for (Persona p: beneficiarios) {
			total+=p.calcularPromedio();
			}
		return total/beneficiarios.size();
	}
	
	
	public int cuposSobrantes() {
		return this.cupos - beneficiarios.size();
	}
	public ArrayList<Estudiante> getBeneficiarios() {
		return beneficiarios;
	}

	public static ArrayList<Subsidio> getListaSubsidio() {
		return listaSubsidio;
	}

	public static void setListaSubsidio(ArrayList<Subsidio> listaSubsidio) {
		Subsidio.listaSubsidio = listaSubsidio;
	}

	public String getNombre() {
		return nombre;
	}

	public void setNombre(String nombre) {
		this.nombre = nombre;
	}

	public float getCondicion() {
		return condicion;
	}
	
	public static void sortByIES() {
		Collections.sort(Estudiante.getListaEstudiantes(), new SortbyIES());
	}

}

class SortbyIES implements Comparator<Estudiante> {

	// Method
	// Sorting in ascending order of name
	public int compare(Estudiante a, Estudiante b) {

		return (int) (Subsidio.calcularIES(a) - Subsidio.calcularIES(b)) * 1000;
	}
}

File: srcpy/gestorAplicacion/gestion/subsidio.py
import math


class Subsidio:
    listaSubsidio = []

    def __init__(self, nombre, condicion, cupos):

        from gestorAplicacion.personas.estudiante import Estudiante
        from gestorAplicacion.personas.profesor import Profesor

        self.nombre = nombre
        self.condicion = condicion
        self.cupos = cupos
        Subsidio.listaSubsidio.append(self)
        Estudiante.getListaEstudiantes().sort()
        Profesor.getListaProfesores().sort()
        self.setBeneficiarios()

    @classmethod
    def calcularIES(cls, beneficiario):
        from gestorAplicacion.personas.estudiante import Estudiante

        puntuacion = (beneficiario.getIngresosFamiliares() * 100) / (
            beneficiario.getNumeroFamiliares() * 760000
        )
        ubicacion = math.pow(beneficiario.getEstrato(), 2) - 4
        if beneficiario.isColegioPublico():
            puntuacion -= 10
        puntuacion += ubicacion - 100
        return puntuacion

    def setBeneficiarios(self):
        from gestorAplicacion.personas.estudiante import Estudiante

        estudiantesGeneral = Estudiante.getListaEstudiantes()
        for estudiante in estudiantesGeneral:

            result = Subsidio.calcularIES(estudiante)
            cuposDisponibles = self.cupos > 0
            if result < self.condicion and cuposDisponibles:
                self.beneficiarios.append(estudiante)
                self.cupos -= 1

    def promedioBeneficiarios(self):
        from gestorAplicacion.personas.estudiante import Estudiante

        total = 0
        for persona in self.beneficiarios:
            total += persona.calcularPromedio()
        return total / len(self.beneficiarios)

    def cuposSobrantes(self):
        return self.cupos - len(self.beneficiarios)

    def getBeneficiarios(self):
        return self.beneficiarios

    @classmethod
    def getListaSubsidio(cls):
        return cls.listaSubsidio

    @classmethod
    def setListaSubsidio(cls, lista):
        cls.listaSubsidio = lista

    ##Metodo de comparacion para estudiantes

    def getNombre(self):
        return self.nombre

    def setNombre(self, nombre):
        self.nombre = nombre

    def getCondicion(self):
        return self.condicion
